File: .vscode/Practise  c programs/circular_queue.c
#include <stdio.h>
int queue[5];
int front = -1;
int rear = -1;
void enqueue(int x)
{
    if (front == -1 && rear == -1)
    {
        front = rear = 0;
        queue[rear] = x;
    }
    else if (((rear + 1) % 5) == front)
    {
        printf("Queue if full\n");
    }
    else
    {
        rear = (rear + 1) % 5;
        queue[rear] = x;
    }
}
void dequeue()
{
    if (front == -1 && rear == -1)
    {
        printf("Queue is empty\n");
    }
    else if (front == rear)
    {
        front = rear = -1;
    }
    else
    {
        printf("%d ", queue[front]);
        front = (front + 1) % 5;
    }
}
void display()
{
    int i = front;
    if (front == -1 && rear == -1)
    {
        printf("Queue is empty\n");
    }
    else
    {
        printf("Queue is %d ", queue[i]);
        i = (i + 1) % 5;
    }
    printf("%d ", queue[rear]);
}
int main()
{
    enqueue(12);
    enqueue(23);
    enqueue(34);
    enqueue(45);
    enqueue(56);
    dequeue();
    display();
    return 0;
}

File: 21 - Queues/CircularQueue.java
import java.util.*;
public class CircularQueue
{
    static class Queue 
    {
        static int arr[];
        static int size;
        static int rear;
        static int front;
        Queue(int n)
        {
            arr=new int[n];
            size=n;
            rear=-1;
            front=-1;
        }
        public static boolean isEmpty()
        {
            return rear==-1 && front==-1;
        }
        public static boolean isFull()
        {
            return (rear+1)%size==front;
        }
        public static void enqueue(int data)
        {
            if(isFull())
            {
                System.out.println("Overflow");
                return;
            }
            //? add 1st element 
            if(front==-1)
            {
                front=0;
            }
            rear=(rear+1)%size;
            arr[rear]=data;
        }
        public static int dequeue()
        {
            if(isEmpty())
            {
                System.out.println("Underflow");
                return -1;
            }
            int result=arr[front];
            //? delete last
            if(rear==front)
            {
                rear=front=-1;
            }
            else
            {
                front=(front+1)%size;
            }
            return result;
        }
        public static int peek()
        {
            if(isEmpty())
            {
                System.out.println("Underflow");
                return -1;
            }
            return arr[front];
        }
    }
    public static void main(String args[])
    {
        Queue q=new Queue(3);
        q.enqueue(1);
        q.enqueue(2);
        q.enqueue(3);
        //? 1 2 3
        while(!q.isEmpty())
        {
            System.out.println(q.peek());
            q.dequeue();
        }
    }
}